_blog/2023-11-15-ssg-in-rust.mdx

```diff
@@ -0,0 +1,355 @@
+
+I created a `Shuttle.toml` file and added the following:
+```rust
+// Shuttle.toml
+assets = ["templates/*", "templates/**/*"]
+```
+This lets Shuttle know where we keep all of our files so that when we deploy, it should just include the files automatically.
+
+Anyway, onto the next part!
+
+## 59:39 - Adding basic functionality
+I knew initially from previous experience that you can use `pulldown-cmark` to turn Markdown to HTML, so I quickly added it using `cargo add pulldown-cmark`. I remember that they had an example on their GitHub repo for parsing Markdown, so I grabbed it and quickly threw it in a handler function:
+```rust
+// src/main.rs
+async fn parser(State(state): State<AppState>, Path(page): Path<String>) -> impl IntoResponse {
+    let mut file = format!("templates/{page}.md");
+
+    let markdown_input = match tokio::fs::read(file).await {
+        Ok(res) => res,
+        Err(_) => return Html("Couldn't find this page - does it exist?".to_string()),
+    };
+
+    let string_output = std::str::from_utf8(&markdown_input).unwrap();
+
+    let mut html_output = String::new();
+    let parser = pulldown_cmark::Parser::new(&string_output);
+    
+    pulldown_cmark::html::push_html(&mut html_output, parser);
+
+    Html(html_output)
+
+```
+
+I made the handler generic by using the Path extractor, then quickly made the `templates` folder and put in a `hello_world.md` file:
+```markdown
+## Hello world!
+```
+
+I then used `cargo clippy` to make sure it all works. There's no errors. After that I used `cargo shuttle run` to get it working and then made the mistake of going to the base route at `/` instead of at `/hello_world` where the web service could read the Markdown file. Never mind! I'll fix that later - it looks like going to `/hello_world` works. Unfortunately, this part took the better part of 10 minutes due to waiting for Rust to compile; granted however, I'm on a laptop with 4GB of RAM. We're not exactly expecting NASA-level performance out of this one.
+
+## 47:12 - Adding CSS styling
+Now that we've got basic markdown parsing functionality, we can move to the next part: styling our pages. I didn't want to spend a lot of time on this - if you're not careful, styling can eat up quite a lot of your time! I primarily wanted to just make sure the page didn't look horrible, so I added a small amount of CSS to center-align the text:
+```css
+// templates/styles.css
+html {
+    display: flex;
+    text-align: center;
+    flex-direction: column;
+    align-items: center;
+    justify-content: center;
+    width: 90%;
+}
+```
+As you can see, it's not quite Awwwards-worthy CSS. However, it'll do for making our pages look at least somewhat aligned.
```

Comment:
typo or intended?

Reply:
Oh, maybe I should have a link to what Awwwards is there.

Awwwards is a website for showcasing the best looking websites and they have competitions regularly